hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java

```diff
@@ -100,15 +124,19 @@ public void increaseSizeLeaving(DatanodeDetails dui, long size) {
    *
    * @return the nex candidate source data node.
    */
-  @Override
-  public DatanodeDetails getNextCandidateSourceDataNode() {
+  private DatanodeDetails getNextCandidateSourceDataNode() {
     if (potentialSources.isEmpty()) {
       LOG.info("no more candidate source data node");
       return null;
     }
     return potentialSources.poll().getDatanodeDetails();
```

Comment:
Unrelated to this PR. Since we are doing a poll here, the source datanode will not be considered again in the iteration?

Reply:
Can you please check if this is an issue? We can create a new jira to address it if it is.

Reply:
this is not an issue. this happens when i tried to refactor the findTargetStrategy and remove ContainerBalancerSelectionCriteria. i have canceled this refactor, so `getNextCandidateSourceDataNode ` is public now

Reply:
The issue is that we are doing a poll in potentialSources. As a result source datanode will be removed.

Reply:
I think this source is added back [here](https://github.com/apache/ozone/blob/e01b47184818365dcb7baf528fb2d874209b6b47/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/FindSourceGreedy.java#L90), and that method is called [here](https://github.com/apache/ozone/blob/e01b47184818365dcb7baf528fb2d874209b6b47/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java#L619). It seems correct but the logic is a bit confusing (https://github.com/apache/ozone/pull/2808#discussion_r773675544)

Reply:
@lokeshj1703 @siddhantsangwan 
```
  private void incSizeSelectedForMoving(DatanodeDetails source,
                                        ContainerMoveSelection moveSelection) {
    ````````
    // update sizeLeavingNode map with the recent moveSelection
    findSourceStrategy.increaseSizeLeaving(source, size);

    // update sizeEnteringNode map with the recent moveSelection
    findTargetStrategy.increaseSizeEntering(target, size);
  }
```
```
 FindSourceGreedy#increaseSizeLeaving

  public void increaseSizeLeaving(DatanodeDetails dui, long size) {
    Long currentSize = sizeLeavingNode.get(dui);
    if(currentSize != null) {
      sizeLeavingNode.put(dui, currentSize + size);
      //reorder according to the latest sizeLeavingNode
      potentialSources.add(nodeManager.getUsageInfo(dui));
      return;
    }
    LOG.warn("Cannot find datanode {} in candidate source datanodes",
        dui.getUuid());
  }
```

if we can find a target and a container for this source datanode, then `incSizeSelectedForMoving` will be definitely called , so this source data node will be added back to that priority queue and re-sort again.

if we can not find any target and container for this source datanode , what we should do is just removing it. 

so i don`t think this is an issue. does this explanation make sense to you?